### ISSUE TITLE
[cssom-view] Consider adding a programmatic way of resetting the visual viewport scale.

### ISSUE THREAD

**Issue by emilio** [w3c/csswg-drafts#9787]
I just came across https://bugzilla.mozilla.org/show_bug.cgi?id=1873934, which seems like a Firefox bug, but contains an interesting code snippet. Simplified below would be:

```
function resetZoom() {
  document.head.querySelector("meta[name=viewport]")
    .setAttribute("content", "width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=1.0, user-scalable=yes")
}
```

As hacky as it looks, the use case seems reasonable to me:

> This is a map application. As such, pinching out the main screen will zoom in the map.
However, when selecting a pin, the popup appears and will open full screen. Pinching out there will zoom in the text field - so far so good.
>
> However, when closing the popup, we want to reset the user scaling to 1.0. Otherwise, some of the controls might be out of view and might not be reachable, as panning will pan the map, not the view.

Given you can pretty much achieve any scaling with the "change the meta viewport dynamically", it might be worth making the `VisualViewport.scale` property writable by authors?

Or if not, at least provide a way of resetting it programmatically?

thoughts? @bokand @theres-waldo @hiikezoe @smfr

**Comment by emilio** [comment]
This would introduce the capability of programmatically (re)set the scale on desktop too (which supports pinch zooming, but doesn't support `<meta name=viewport>`). That seems reasonable to me but I thought I should point that out.

If we just provide an API to reset it, we don't need to think too hard, but if we just set the scale to be writable then there are some other questions about whether we should limit the API range to some sane values or what not. Browsers seem to have some maximum range they allow zooming to, not sure how interoperable that is. Also in desktop you can't zoom out.

That makes me think that an explicit "reset the scale" API might be better, but I don't have a super-strong opinion.

**Comment by pietervdvn** [comment]
As requested, this is the stack overflow post that inspired the hack: https://stackoverflow.com/questions/13162174/how-to-enable-user-scalable-viewport-property-after-disabling-it/16933635#16933635

**Comment by theres-waldo** [comment]
My recollection from past discussions about the visual viewport is that it was intended to be an "inert" feature, i.e. a way for a user to control their view of the page with no or minimal effect on the content being viewed.

The Visual Viewport API has eroded that a bit, since sites can now react to visual viewport movements, but allowing sites to set the pinch-zoom scale could open the door to new user-hostile behaviours, such as trying to prevent the user from zooming in at all, or from viewing a certain part of the page zoomed in, etc.

That said, on Android at least, we have an accessibility setting called "Zoom on all websites" which overrides current ways for a page to disable pinch-zooming (such as `user-scalable=no`). Perhaps allowing the page to modify the scale could be fine as long as this override behaviour applied to that as well?

---

Regarding the use case discussed in the bug and the StackOverflow question, I wonder if it points to something else we could be doing better: giving the page a more explicit way to say "the page is now handling zooming, the browser shouldn't handle it". This came up in [bug 1866028](https://bugzilla.mozilla.org/show_bug.cgi?id=1866028), where such a mechanism would avoid situations where users accidentally trigger browser pinch-zoom on pages that intended to handle the gesture themselves, leading to e.g. controls being stuck offscreen (outside of the visual viewport).

If we had such an API, we could then e.g. say that whenever the page is handling zooming, the browser's zoom level should be 1, thereby getting this "reset" behaviour automatically.

**Comment by emilio** [comment]
> The Visual Viewport API has eroded that a bit, since sites can now react to visual viewport movements, but allowing sites to set the pinch-zoom scale could open the door to new user-hostile behaviours, such as trying to prevent the user from zooming in at all, or from viewing a certain part of the page zoomed in, etc.

Isn't that already preventable effectively via the viewport meta tag (modulo the feature you mentioned earlier)? Or by transforming the content by an inverse of the visual viewport scale?

**Comment by smfr** [comment]
I really don't like the idea of pages being able to override the user's zoomed in state, especially since zooming in is often done as a mitigation for poor text readability. It would be the equivalent of pages force-scrolling you to the top. Sure, it's possible on mobile devices by messing with the viewport tag, but let's not propagate this to desktop.

**Comment by theres-waldo** [comment]
> > The Visual Viewport API has eroded that a bit, since sites can now react to visual viewport movements, but allowing sites to set the pinch-zoom scale could open the door to new user-hostile behaviours, such as trying to prevent the user from zooming in at all, or from viewing a certain part of the page zoomed in, etc.
> 
> Isn't that already preventable effectively via the viewport meta tag (modulo the feature you mentioned earlier)? Or by transforming the content by an inverse of the visual viewport scale?

Yeah, in that sense this proposal would just make it easier for sites to misbehave in ways they already technically could.

---

Another consideration that has occurred to me is implementation complexity. In Firefox, the fact that the scroll position can be modified in both the content process (by web content) and the compositor process (by user input events), and these two pieces of state with independent sources of updates to them therefore need to be reconciled with each other, is a significant source of complexity and bugs. Making the zoom level writable in the content process would carry over that complexity to the zoom level as well, especially if there is an expectation that the modified value of `scale` is immediately reflected when reading back `scale`.

(If the API is a `reset()` function whose semantics are that it resets the `scale` to 1 but "in the near future" rather than "immediately", such than an immediate readback of `scale` may not yet reflect the change, this concern would be largely addressed.)

**Comment by bokand** [comment]
I'll echo the worry that this is more likely to be misused than helpful - though a reset mechanism seems less risky.

Given that the app is already handling zoom itself, it seems like the "right" fix to me is prevent zooming over the popup/text field (e.g. with `touch-action`). Blocking zoom isn't great in general but the app is already handling zoom in part of its UI itself - mixing and matching app-handled and browser-handled zoom seems fraught.
